### DISCUSSION TITLE
export function auth<T>(connectionString: string, middleware: Middleware<[T]>){...}

### DISCUSSION THREAD

**Discussion by HenriqueStocco** [thtsystems/open-visit#16] [Q&A] [answered]
I'd like to know what the <T> stands for and what it's for?

**Comment by ernestoresende** [comment, marked as answer]
I'm assuming you're referring to this section here, on our `auth` workspace, which initializes the Lucia configuration needed to handle authentication tasks.

https://github.com/thtsystems/open-visit/blob/ee9e868613200d751f24d2ae8d0849683a826b50/auth/auth.ts#L14-L34

The `<T>` portion of this is a TypeScript Generic. It's kinda hard to explain without more concrete examples, but think of it as a dynamic function parameter, but for types. The `auth` function from Lucia expects a `middleware` property, which can be one of any of the exports from `lucia/middleware`. As Lucia's own documentation states:

> However, every framework and runtime has their own representation of an incoming request and outgoing response, such as the web standard `Request/Response` and Node.js’ `IncomingMessage/OutgoingMessage`.

Hono, for example, has the `context` parameter for each function handler, which is built on top of the web standard `Request/Response`. Other backend frameworks have other implementations, and so on...

With all that said, the reason we use this generic here, instead of just exporting Lucia's `lucia()` (eg.: `export const auth = lucia({...})`), which would also work just fine, is that by instantiating it as a function, we can dynamically pass in a middleware as a parameter, and the type will be automatically picked up by TypeScript. We don't need to create a whole new `lucia()` function call and assign it to a new variable just to pass a different middleware.

The topic of generics is fairly complex, specially if you're not used to statically typed languages, so it's fine to not completely grasp the concept of it immediately. I'll leave some resources that might help:

- How To Use Generics in TypeScript: https://www.digitalocean.com/community/tutorials/how-to-use-generics-in-typescript 
- The Middleware API definition from Lucia's docs: https://lucia-auth.com/reference/middleware/#middleware

**Reply by HenriqueStocco** [reply]
Thank you very much for that reply, I will read the material to understand it a little better. I was in doubt because in the typescript documentation, from what I understand, this <T> reserves the type of the parameter passed when instantiated, and then it would use it for something. Well, I'm sorry for the bad question too, I should have put the code, the path and asked the question afterwards.

**Reply by ernestoresende** [reply]
No need to worry about that, all questions are welcome.

> from what I understand, this reserves the type of the parameter passed when instantiated, and then it would use it for something

This is a very good way to put it, yes. This is the tooling behind so much of type inference on libraries that heavily rely on TypeScript. It uses type information about what you pass in to build dynamic types that can change based on input.

**Reply by HenriqueStocco** [reply]
Typescript is a really complex tool

**Comment by HenriqueStocco** [comment]
Another question, if I upload the `compose`, should I have the tables? And I want to make a `CRUD` of each path, but when I look at the tables (I use `dbeaver`), there are no tables, even importing the schema in the code, I don't understand this, could you please explain? I know I was supposed to wait for the issues, but my anxiety wouldn't let me lol

**Reply by ernestoresende** [reply]
Took me a while to see this, but since this comment was created, most of the things we did so far must answer it. See:
- #42 

And also review the contribution guide, along with the newly opened issues for the CRUD operations. Feel free to open a new issue if you have any questions regarding project setup or how to author the API routes.
